### ISSUE TITLE
Showing validation errors on collapsed sections

### ISSUE THREAD

**Issue by lelit** [wagtail/wagtail#8990]
### Is your proposal related to a problem?

I quickly tried the [new 4.0 editor](https://docs.wagtail.org/en/latest/releases/4.0.html#page-editor-redesign) and I found it very nice, kudos!

However there is a thing that I find very difficult, in particular with the new default of collapsing all sections in the edit page: when there are validation errors, there's no way to quickly understand *which* section(s) contains the problematic fields. I'm force to open each section and scroll all the fields, until I find those with the red error.

### Describe the solution you'd like

Maybe the affected sections could be marked in some way, for example rendering them in red, or even put a badge with the number of errors?

**Comment by lelit** [comment]
Another option could be make the indicator at the page top a button, that opens/scrolls to the first erroneous field.

**Comment by thibaudcolas** [comment]
Hey @lelit, thanks for the report. Could you share a screenshot of what this looks like for you? 

> in particular with the new default of collapsing all sections in the edit page

This is the bit I don’t understand – we don’t collapse anything by default. And even when a site implementer marks a section as "collapsed by default", we display it expanded if there are errors within it.

**Comment by lelit** [comment]
I'm **really** sorry for being inaccurate, as I didn't mention that I'm using `StreamField`s and in the specific case I have an home page with
```python
    body = StreamField(
        HomePageBlock(),
        blank=True,
        collapsed=True
    )
```
where `HomePageBlock` is something like
```python
class HomePageBlock(StreamBlock):
    presentation = PresentationBlock()
    about = AboutBlock()
    service = ServiceBlock()
    ...
```

At that point, when one of the blocks has errors, I see the following:
![example1](https://user-images.githubusercontent.com/412797/184845490-00377b36-73b2-4dc1-b2c3-053af99c0c03.png)

To discover where the error is I have to open each section in turn, until I find it:
![example2](https://user-images.githubusercontent.com/412797/184845810-5105d206-5bca-478d-849d-b2dec6dda99e.png)

Thank you for your attention!

**Comment by lb-** [comment]
Is this a duplicate of https://github.com/wagtail/wagtail/issues/4122 ?

**Comment by gasman** [comment]
@lb- No, I don't think so. #4122 is about StreamField not handling custom validation errors that are triggered outside of StreamField's own mechanism; this one is about the interaction between validation errors (including ones coming from StreamField itself) and the collapse feature.

**Comment by thibaudcolas** [comment]
Thank you for the details @lelit! From what you’re describing, this seems to be an issue with StreamField blocks’ collapsed state. I suspect this was an issue even in previous releases, but won’t have the chance to confirm this or investigate further for the time being (I can get back to it in late August if no one else looks into it earlier).

The logic to not-collapse sections with errors for everything but StreamField is here: https://github.com/wagtail/wagtail/blob/main/client/src/includes/panels.ts#L53-L59

For StreamField, I’m less familiar with its implementation but the corresponding code seems to be here: https://github.com/wagtail/wagtail/blob/main/client/src/components/StreamField/blocks/BaseSequenceBlock.js#L335. This doesn’t seem to have a check for errors in the collapsing logic, though it could be that the error handling expands collapsed blocks subsequently in the blocks’ initialisation, or we mistakenly removed this as part of recent changes.

**Comment by lelit** [comment]
Yes, I can confirm that the same happens on Wagtail 3.x.

**Comment by thibaudcolas** [comment]
👍 thank you @lelit. I’ve recently worked on this and am pretty sure this just never got implemented. 

Here is our logic that keeps sections from collapsing if there is an error: https://github.com/wagtail/wagtail/blob/e9ce02d40939931e2d9df05855c3a185d3095386/client/src/includes/panels.ts#L66-L74

And the equivalent code that collapses StreamField blocks: https://github.com/wagtail/wagtail/blob/e9ce02d40939931e2d9df05855c3a185d3095386/client/src/components/StreamField/blocks/BaseSequenceBlock.js#L235-L237

So – we’d need to similarly update this code to query for the presence of errors within the block. I think doing this with DOM queries (`querySelector( '[aria-invalid="true"], .error, .w-field--error');`) could work well considering we want to do this regardless of where in the DOM tree the error is.

I’ll tag this as a `good first issue` – there is a fair bit of work needed to reproduce the problem and fix it, but I think the solution seems much clearer now.

**Comment by lelit** [comment]
Thanks for the details, I may try to implement the above, if/when time permits.
Is there a "quick" intro to setup a development environment, in particular wrt TS compilation?

**Comment by salty-ivy** [comment]
@lelit are you still working on the issue? can i pick it up?

**Comment by lelit** [comment]
No, I couldn't dedicate any time yet, sorry. So, go for it, thank you!

**Comment by salty-ivy** [comment]
Do we actually need the collapsing logic in contrast validation errors in collapsed section?
wouldn't this side vertical navigation panel is serving what this issue is asking for ?
![image](https://user-images.githubusercontent.com/74553951/200874480-8d9c2633-767a-4ad4-9a63-fbf7d758f7dc.png)

**Comment by salty-ivy** [comment]
> @lelit  Do we actually need the collapsing logic in contrast to validation errors in collapsed section?
> 
> wouldn't this side vertical navigation panel is serving what this issue is asking for ?
> 
> ![image](https://user-images.githubusercontent.com/74553951/200874480-8d9c2633-767a-4ad4-9a63-fbf7d758f7dc.png)
> 
> 
> 
> EDIT: my bad i misunderstood the issue , i see the issue is finding errors within the stream field
> 
> but still the vertical bar pretty much navigates you to that perticular black that has error in it.
> 
>  

@lelit could you please confirm
The comment here? Just wanna understand the issue completely

**Comment by salty-ivy** [comment]
![image](https://user-images.githubusercontent.com/74553951/201008859-c8280740-097b-4342-81e4-3bf6e58e1f6d.png)
as you can see its clearly navigating us to the errored block 
@lelit  @thibaudcolas

**Comment by lelit** [comment]
It's nice that the UI gets better so fast! :wink: Many thanks to everybody's involved!
I didn't have the occasion to try out that enhancement, that surely looks pretty, but I'm not in position to judge it as _good enough_ right now, sorry.
I'll do my best to try it out in the following days, so to be able to give a more informed opinion.

**Comment by salty-ivy** [comment]
> It's nice that the UI gets better so fast! 😉 Many thanks to everybody's involved! I didn't have the occasion to try out that enhancement, that surely looks pretty, but I'm not in position to judge it as _good enough_ right now, sorry. I'll do my best to try it out in the following days, so to be able to give a more informed opinion.

surely try it out at your ease, and see if it serves the purpose of this issue, then we can close  this issue on that basis

**Comment by thibaudcolas** [comment]
I don’t think the new minimap component should be considered as part of this issue. Please re-review my comment at https://github.com/wagtail/wagtail/issues/8990#issuecomment-1279536431 – some parts of the form will reveal themselves when there are errors within, there’s no reason we wouldn’t want that for all parts of the form.

**Comment by salty-ivy** [comment]
Well sure ahead just wanted to confirm it out, in that case I shall begin working on it soon

**Comment by TonisPiip** [comment]
Would it be possible to have a quick fix to show which section has the error using the has() css selector?  :thinking: 

has() doesn't have full browser support, but it's something anyone can add, even when using older versions of TW.

**Comment by lelit** [comment]
I did a quick upgrade of one of my apps to Wagtail 4.1.1, just enough to have a better idea of the new edit UI.

The vertical navigation panel is nice and handy, but at least with my configuration (remember, I'm using the new default of collapsing all sections) is not ideal: when I have an error in one of the section, I see the red badge in the side navigation, but clicking on it has basically no effects on the form, because being all sections collapsed they fits in the browser window and thus nothing scrolls.

I think a better experience would be:

1. automatically expand sections containing errors (as suggested by @thibaudcolas), and/or
2. clicking on the navigation panel links expands and brings to visible the related section

**Comment by TonisPiip** [comment]
Maybe some change to how collapsing is handled in the FE should be changed. Haven't looked it it. But it would also be good if it was somewhat statefull. Its annoying when saving a draft, and having everything close.

If it state of the what is open is saved somewhere (localsore or just in the html itself), then it also shouldn't be *too* hard to have the logic checking for if something is open also depend on if there's any errors.

**Comment by ArijeetBanerjee** [comment]
I want to work on this issue. kindly give me permission and access to work on this issue

**Comment by lb-** [comment]
Note: https://github.com/wagtail/wagtail/issues/10090 has been raised, it may be a simpler starting point before doing this ticket as it is like for like migration from ad-hoc JS to Stimulus.

Once the 'count controller' has been implemented, it should not be too hard to use the same set of data attributes on panels to show counts there.

**Comment by Stormheg** [comment]
I've created a PR with a fix for StreamFields here: https://github.com/wagtail/wagtail/pull/10673

**Comment by gasman** [comment]
(reopening as #10673 was only a partial fix)
